### PR TITLE
Internal Gates: Adds memory allocation gates to check-ins

### PR DESCRIPTION
**Pull request by j82w**
# Pull Request Template

## Description

This PR ensures that if the memory allocation change by over 5% that gates will fail and the baseline will need to be updated. This will prevent accidental regressions for allocations. This is done by running the performance tests 3 times which has the MemoryDiagnoser from BenchmarkDotnet enabled and averages the results of those runs. It then is compares the results to the baseline file and fails if the it varies by more than 5%. 

This PR will not support latency validation as the gates have to much variance in performance to support reliable validation.

## Type of change

Please delete options that are not relevant.

- [] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

## Closing issues

To automatically close an issue: closes #IssueNumber

### CODE REVIEW

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Contracts/PerformanceValidation.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+                Console.WriteLine("CurrentBenchmarkResults count does not match the baseline BenchmarkResults.json. Please update the BenchmarkResults.json: " + currentBenchmarkResults);
+                return 1;
+            }
+
+            foreach(KeyValuePair<string, double> currentResult in operationToMemoryAllocated)
+            {
+                double baselineResult = baselineBenchmarkResults[currentResult.Key];
+
+                // Add 1000 bytes of buffer to avoid minor variation between test runs
+                double diff = currentResult.Value - baselineResult;
```
Comment (bchong95) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Contracts/PerformanceValidation.cs:69]: absolute value differences don't make sense when some benchmarks use more memory than others.

**Review by bchong95 — CHANGES_REQUESTED** [review]

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Contracts/PerformanceValidation.cs (hunk 1 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Contracts/PerformanceValidation.cs:67]: Changed it to be %5 of the baseline results

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Contracts/PerformanceValidation.cs [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        {
+            // Using dotnet run in the gates puts the current directory at the root of the github project rather than the execute folder.
+            string currentDirectory = Directory.GetCurrentDirectory();
+            int removePathsLowerThanIndex = currentDirectory.IndexOf(@"\Microsoft.Azure.Cosmos");
+            if(removePathsLowerThanIndex >= 0)
+            {
+                currentDirectory = currentDirectory.Remove(removePathsLowerThanIndex);
+            }
+            
+            currentDirectory += @"\Microsoft.Azure.Cosmos\tests\Microsoft.Azure.Cosmos.Performance.Tests\bin\Release\netcoreapp3.1\Contracts\";
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Contracts/PerformanceValidation.cs:60]: If someone runs it locally without Release configuration, would this work? Should we have a compilation IF for the paths?

**Review by ealsur — COMMENTED** [review]
Is the expectation that users run it locally in their machine to update the baseline and submit an updated JSON baseline?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Contracts/PerformanceValidation.cs [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+            {
+                double baselineResult = baselineBenchmarkResults[currentResult.Key];
+
+                // Add 5% buffer to avoid minor variation between test runs
+                double diff = currentResult.Value - baselineResult;
+                double maxAllowedDiff = baselineResult * .05;
+                if (diff > maxAllowedDiff)
+                {
+                    Console.WriteLine("The current results have exceed the baseline memory allocations. Please fix the performance regression. " +
+                        "If this is by design please update the BenchmarkResults.json file using the CurrentBenchmarkResults.json in the output folder: " + currentBenchmarkResults);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Contracts/PerformanceValidation.cs:85]: Can we indicate how should the local user run the validation/update on the error message?

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Contracts/PerformanceValidation.cs (hunk 2 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Contracts/PerformanceValidation.cs]: Fixed

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Contracts/PerformanceValidation.cs (hunk 3 above)
Comment (j82w) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Performance.Tests/Contracts/PerformanceValidation.cs]: Please check the latest message

**Comment by j82w** [comment]
> Is the expectation that users run it locally in their machine to update the baseline and submit an updated JSON baseline?

Yes, they can run it locally on their machine or they can get the output from the test in the gate to update the JSON baseline file. It works the same as the contract validation does.

**Review by ealsur — APPROVED** [review]

**Review by kirankumarkolli — DISMISSED** [review]

**Review by bchong95 — APPROVED** [review]

**Review by kirankumarkolli — APPROVED** [review]

**Comment by kirankumarkolli** [comment]
Unrelated test failure 
image

**Pull request merged**